### PR TITLE
Rebase support

### PR DESCRIPTION
**Pull request by bpeabody**
I began the work in this PR to better support rebasing; and in particular to add `rebase ---continue`, and `rebase --abort`. However, in the process I uncovered a bug that had been introduced when we switched to using detached heads in submodules caused by https://github.com/twosigma/git-meta/issues/115.

The problem was detected only after I started reading the rebase state of a repository and my existing rebase unit tests detected lingering rebase information in submodules after successful completion of a rebase.

I am submitting these changes now, before finishing `rebase --abort` and `rebase --continue` so that I can fix this bug; the fix and test for it depend on the other work in this PR.

### CODE REVIEW

File: node/lib/util/rebase_file_util.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+ * behavior is undefined unless `gitDir` is a directory.
+ *
+ * @param {String} gitDir
+ * @return {String|null}
+ */
+exports.findRebasingDir = co.wrap(function *(gitDir) {
+    assert.isString(gitDir);
+    const files = yield fs.readdir(gitDir);
+    const rebaseCheck = /^rebase-..*/;
+    for (let i = 0; i < files.length; ++i) {
```
Comment (novalis) [inline node/lib/util/rebase_file_util.js]: Just check for the known ones: rebase-apply and rebase-merge.

File: node/lib/util/rebase_file_util.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    let headName;
+    let originalHead;
+    let onto;
+
+    // Try to read old head 
+    try {
+        headName = yield fs.readFile(path.join(rebasingDir,
+                                               exports.headFileName),
+                                     "utf-8");
+        originalHead  = yield fs.readFile(path.join(rebasingDir, "orig-head"),
```
Comment (novalis) [inline node/lib/util/rebase_file_util.js]: nit: extra space before = (also on line 112)

File: node/lib/util/write_repo_ast_util.js [hunk 3]
```diff
@@ -399,6 +400,24 @@ git checkout-index -a -f
...
+    // set up a rebase if there is one
+
+    if (null !== ast.rebase) {
+        const rebase = ast.rebase;
+        const originalSha = commitMap[rebase.originalHead];
+        const ontoSha = commitMap[rebase.onto];
+        const original = yield NodeGit.AnnotatedCommit.lookup(repo,
+                                                              originalSha);
+        const onto = yield NodeGit.AnnotatedCommit.lookup(repo, ontoSha);
+        yield NodeGit.Rebase.init(repo, original, onto, null, null);
```
Comment (novalis) [inline node/lib/util/write_repo_ast_util.js]: Something seems weird here -- the nodegit docs say that the args to Rebase.init are (repo, branch, upstream, onto, opts).

File: node/test/util/write_repo_ast_util.js [hunk 4]
```diff
@@ -159,12 +166,25 @@ describe("WriteRepoASTUtil", function () {
...
+                expected: "S:C2-1;C3-1;Bfoo=3;Bmaster=2;Efoo,2,3;H=3",
+            },
         };
 
         Object.keys(cases).forEach(caseName => {
-            const shorthand = cases[caseName];
+            let input = cases[caseName];
+            let expected = input;
+
+            // Sometimes, the expected read value is not the same as the write
```
Comment (novalis) [inline node/test/util/write_repo_ast_util.js]: This comment makes no sense to me.

File: node/lib/util/repo_ast_util.js [hunk 5]
```diff
@@ -436,6 +436,31 @@ ${colorExp(expected.currentBranchName)}`
...
+    if (null === actual.rebase && null !== expected.rebase) {
+        result.push("Missing rebase.");
+    }
+    else if (null !== actual.rebase && null === expected.rebase) {
+        result.push("Unexpected rebase.");
+    }
+    else if (null !== actual.rebase) {
+        if (actual.rebase.headName !== expected.rebase.headName) {
+            result.push(`Expected ${colorBad("rebase head name")} to be \
+${colorExp(actual.rebase.headName)} but got \
```
Comment (bpeabody) [inline node/lib/util/repo_ast_util.js]: yep, looks like I repeated this mistaken pattern several times

File: node/lib/util/rebase_file_util.js (hunk 2 above)
Comment (bpeabody) [inline node/lib/util/rebase_file_util.js]: good eye

File: node/lib/util/write_repo_ast_util.js (hunk 3 above)
Comment (bpeabody) [inline node/lib/util/write_repo_ast_util.js]: That method has a weird, overloaded, contract. The documentation for `branch` is: "The terminal commit to rebase, or NULL to rebase the current branch", and for `upstream`: "The commit to begin rebasing from, or NULL to rebase all reachable commits".

File: node/test/util/write_repo_ast_util.js (hunk 4 above)
Comment (bpeabody) [inline node/test/util/write_repo_ast_util.js]: took another stab at it

**Comment by bpeabody** [comment]
I think I addressed your requests. I'm not sure why it still says there is a conflict; I just rebased it on top of master.

File: node/lib/util/write_repo_ast_util.js (hunk 3 above)
Comment (novalis) [inline node/lib/util/write_repo_ast_util.js]: Yeah, but here, you're passing "original" for branch and "onto" for "upstream" -- that seems wrong.

Comment (bpeabody) [inline node/lib/util/write_repo_ast_util.js]: I agree, but I think the documentation and/or contract for `NodeGit.Rebase.init` is the source of confusion. The names of the files for the original and onto commits in rebase-apply/merge file are `origHead` and `onto`, respectively.

File: node/lib/util/rebase.js [hunk 6]
```diff
@@ -31,449 +31,55 @@
...
     /**
-     * Begin a rebase for this submodule from the current HEAD onto the
-     * specified `remoteCommitId`.  If a rebase is in progress do nothing.
+     * Create a new `Rebase` object.
      *
-     * @async
-     * @private
-     * @param {String} remoteCommitSha
+     * @param {String} headName
+     * @param {String} orignalHead
```
Comment (novalis) [inline node/lib/util/rebase.js]: spelling on orignalHead.

File: node/test/util/status.js [hunk 7]
```diff
@@ -900,7 +935,11 @@ describe("Status", function () {
         const cases = {
             "trivial": {
                 state: "x=S",
-                regex: /On branch.*master.*\nnothing to commit.*/,
+                regex: /On branch.*master.*\n.*nothing to commit.*/,
+            },
+            "rebase": {
+                state: "x=S:C2-1;C3-1;Bfoo=3;Bmaster=2;Efoo,2,3",
+                regex: /rebas/,
```
Comment (novalis) [inline node/test/util/status.js]: Is this the most specific possible regex?

Comment (bpeabody) [inline node/test/util/status.js]: Really needs to reference commit ids as well; I've got it on my todo list to make all those regexes a little more specific.

**Pull request merged**